Common/Render/Text/draw_text_sdl.cpp

```diff
@@ -0,0 +1,263 @@
+	CacheKey key{ std::string(str, len), fontHash_ };
+
+	TextMeasureEntry *entry;
+	auto iter = sizeCache_.find(key);
+	if (iter != sizeCache_.end()) {
+		entry = iter->second.get();
+	} else {
+		TTF_Font *font = fontMap_.find(fontHash_)->second;
+
+		int width = 0;
+		int height = 0;
+		TTF_SizeUTF8(font, str, &width, &height);
+
+		entry = new TextMeasureEntry();
+		entry->width = width;
+		entry->height = height;
+		sizeCache_[key] = std::unique_ptr<TextMeasureEntry>(entry);
+	}
+
+	entry->lastUsedFrame = frameCount_;
+	*w = entry->width * fontScaleX_ * dpiScale_;
+	*h = entry->height * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerSDL::MeasureStringRect(const char *str, size_t len, const Bounds &bounds, float *w, float *h, int align) {
+	std::string toMeasure = std::string(str, len);
+	int wrap = align & (FLAG_WRAP_TEXT | FLAG_ELLIPSIZE_TEXT);
+	if (wrap) {
+		bool rotated = (align & (ROTATE_90DEG_LEFT | ROTATE_90DEG_RIGHT)) != 0;
+		WrapString(toMeasure, toMeasure.c_str(), rotated ? bounds.h : bounds.w, wrap);
+	}
+
+	TTF_Font* font = fontMap_.find(fontHash_)->second;
+
+	int width = 0;
+	int height = 0;
+	TTF_SizeUTF8(font, str, &width, &height);
+
+	*w = width * fontScaleX_ * dpiScale_;
+	*h = height * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerSDL::DrawString(DrawBuffer &target, const char *str, float x, float y, uint32_t color, int align) {
+	using namespace Draw;
+	if (!strlen(str))
+		return;
+
+	CacheKey key{ std::string(str), fontHash_ };
+	target.Flush(true);
+
+	TextStringEntry *entry;
+
+	auto iter = cache_.find(key);
+	if (iter != cache_.end()) {
+		entry = iter->second.get();
+		entry->lastUsedFrame = frameCount_;
+	} else {
+		DataFormat texFormat = Draw::DataFormat::R4G4B4A4_UNORM_PACK16;
+
+		entry = new TextStringEntry();
+
+		TextureDesc desc{};
+		std::vector<uint8_t> bitmapData;
+		DrawStringBitmap(bitmapData, *entry, texFormat, str, align);
+		desc.initData.push_back(&bitmapData[0]);
+
+		desc.type = TextureType::LINEAR2D;
+		desc.format = texFormat;
+		desc.width = entry->bmWidth;
+		desc.height = entry->bmHeight;
+		desc.depth = 1;
+		desc.mipLevels = 1;
+		desc.tag = "TextDrawer";
+		entry->texture = draw_->CreateTexture(desc);
+		cache_[key] = std::unique_ptr<TextStringEntry>(entry);
+	}
+
+	if (entry->texture) {
+		draw_->BindTexture(0, entry->texture);
+	}
+
+	float w = entry->bmWidth * fontScaleX_ * dpiScale_;
+	float h = entry->bmHeight * fontScaleY_ * dpiScale_;
+	DrawBuffer::DoAlign(align, &x, &y, &w, &h);
+	if (entry->texture) {
+		target.DrawTexRect(x, y, x + w, y + h, 0.0f, 0.0f, 1.0f, 1.0f, color);
+		target.Flush(true);
+	}
+}
+void TextDrawerSDL::DrawStringBitmap(std::vector<uint8_t> &bitmapData, TextStringEntry &entry, Draw::DataFormat texFormat, const char *str, int align) {
+	if (!strlen(str)) {
+		bitmapData.clear();
+		return;
+	}
+
+	std::string processed_str(str);
+	processed_str = std::regex_replace(processed_str, std::regex("&&"), "&");
```

Comment:
we're not using std::regex_replace anywhere else in the code, so pulling it in just for this seems overkill.